---
title: "Evaluation metrics LRGASP -- Challenge 3"
output: html_document
params:
  date: !r Sys.Date()
  output.directory: rdata  #folder where the evaluation output will be saved
  Name: tool_name
  Platform: platform
  platform2: platform2
  library_prep: library_prep
  library_prep2: library_prep2
  data_category: data_category
  data_category2: data_category2
  pipelineCode: pipelineCode
  pipelineCode2: pipelineCode2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
getwd()
```

```{r libraries, echo=FALSE}
## These are a few libraries that must be loaded for the plots and also a couple of variables to make them "nice"x
suppressWarnings(library(ggplot2))
suppressWarnings(library(scales))
suppressWarnings(library(knitr))
suppressWarnings(library(plotly))

myPalette = c("#6BAED6","#FC8D59","#78C679","#EE6A50","#969696","#66C2A4", "goldenrod1", "darksalmon", "#41B6C4","tomato3", "#FE9929")
suppressWarnings(
  mytheme <- theme_classic(base_family = "Helvetica") +
  theme(axis.line.x = element_line(color="black", size = 0.4),
        axis.line.y = element_line(color="black", size = 0.4)) +
  theme(axis.title.x = element_text(size=14),
        axis.text.x  = element_text(size=13),
        axis.title.y = element_text(size=14),
        axis.text.y  = element_text(vjust=0.5, size=13) ) +
  theme(legend.text = element_text(size = 10), legend.title = element_text(size=11), legend.key.size = unit(0.5, "cm")) +
  theme(plot.title = element_text(lineheight=.4, size=13)) +
  theme(plot.margin = unit(c(2.5,1,1,1), "cm")) 

)

```

## Your submission: `r params$Name` on `r params$Platform` (`r params$library_prep` `r params$data_category`) data and `r params$platform2` (`r params$library_prep2` `r params$data_category2`)

### Meta Data Table
```{r table, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
# Create the table data frame
table <- data.frame(
  'Dataset' = c(1,2),
  'Organism' = c(organism, organism),
  'Tool' = c(tool_name, tool_name),
  'Platform' = c(platform, platform2),
  'Library Preparation' = c(library_prep, library_prep2),
  'Data Category' = c(data_category, data_category2)
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

```


### Background

Challenge 3 is evaluated according to a *de novo* long reads-based genome (provided by the LRGASP Consortium) and SIRV Lexogen Set 1 which was included before library preparation.

The LRGASP uses SQANTI3 categories to define evaluating features and metrics for Challenge 3.

### Categories defined by SQANTI3

Due lo the lack of annotation on the *de novo* transcriptome, these categories will only be associated to transcripts that map to SIRV reference sequences. 

- **Full Splice Match (FSM):** Transcripts matching a reference SIRV at all splice junctions.
- **Incomplete Splice Match (ISM):** Transcripts matching consecutive, but not all, splice junctions of the reference SIRV.
- **Novel in Catalog (NIC):**	Transcripts containing new combinations of already annotated splice junctions or novel splice junctions formed from already annotated donors and acceptors. 
- **Novel Not in Catalog (NNC):**	Transcripts using novel donors and/or acceptors.
- **Reference Match (RM):**	FSM transcript with 5´ and 3´ends within 50 nts of the TSS/TTS annotation. This means that a certain SIRV was detected perfectly.

**The rest of the transcripts will be catalogued as Intergenic**

## Evaluation of detected transcripts for Challenge 3  {.tabset .tabset-pills}

These are some definitions used to evaluate the submitted transcriptome:

- **Full Illumina Splice Junction Support:**	Transcripts with all SJ supported by at least one Illumina read.
- **Non-canonical transcripts**	Transcripts with at least one non-canonical junction.
- **Intra-priming:**	It is considered that there is evidence of intra-priming when, in the genomic sequence 20bp downstream the detected TTS, there is at least a 60% of A's.
- **RT-switching:**	Evidence of RT-switching (see SQANTI ref)

## Evaluation Data {.tabset .tabset-pills}

### Quality Control {.tabset .tabset-fade .tabset-dropdown}

#### Global Metrics
```{r global metrics, echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

# Global metrics
kable(temp_env$transcriptome.results["Transcriptome without reference"], align = "r")
```


```{r globel metrics 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

kable(temp_env2$transcriptome.results["Transcriptome without reference"], align = "r")
```

#### Length Distribution

```{r length distr., echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")
# Length distribution
avg_length <- as.numeric(temp_env$transcriptome.results$`Transcriptome without reference`["Average length", "Absolute value"])

if (organism_like == 'mouse'){
  p1 <- ggplot(temp_env$sqanti_data, aes(x=length, fill=structural_category)) +
  geom_histogram(binwidth=100) +
  mytheme +
  scale_fill_manual(values=myPalette) +
  labs(title = "Length distribution",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("Length (bp)") +
  geom_vline(xintercept = avg_length)
} else if(organism_like == 'manatee'){
  p1 <- ggplot(temp_env$sqanti_data, aes(x=length, fill=structural_category)) +
  geom_histogram(binwidth=100) +
  mytheme +
  scale_fill_manual(values=myPalette[1]) +
  labs(title = "Length distribution",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("Length (bp)") +
  geom_vline(xintercept = avg_length)
}


ggplotly(p1)

```

```{r length distr. 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

# Length distribution file 2
avg_length <- as.numeric(temp_env2$transcriptome.results$`Transcriptome without reference`["Average length", "Absolute value"])

if (organism_like == 'mouse'){
  p2 <- ggplot(temp_env2$sqanti_data, aes(x=length, fill=structural_category)) +
  geom_histogram(binwidth=100) +
  mytheme +
  scale_fill_manual(values=myPalette) +
  labs(title = "Length distribution",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("Length (bp)") +
  geom_vline(xintercept = avg_length)
} else if(organism_like == 'manatee'){
  p2 <- ggplot(temp_env2$sqanti_data, aes(x=length, fill=structural_category)) +
  geom_histogram(binwidth=100) +
  mytheme +
  scale_fill_manual(values=myPalette[1]) +
  labs(title = "Length distribution",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("Length (bp)") +
  geom_vline(xintercept = avg_length)
}


ggplotly(p2)
```

#### Minimum SJ Coverage
```{r SJ coverage, echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")
### Minimum SJ coverage

if(organism_like == 'mouse'){
  p1 <- ggplot(temp_env$sqanti_data, aes(x=log(min_cov+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette)+
  labs(title="Minimum SJ coverage",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("log(Minimum coverage of SJ +1)")
} else if (organism_like == 'manatee'){
  p1 <- ggplot(temp_env$sqanti_data, aes(x=log(min_cov+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette[1])+
  labs(title="Minimum SJ coverage",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("log(Minimum coverage of SJ +1)")
}


ggplotly(p1)

```

```{r SJ coverage 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

### Minimum SJ coverage file 2

if(organism_like == 'mouse'){
  p2 <- ggplot(temp_env2$sqanti_data, aes(x=log(min_cov+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette)+
  labs(title="Minimum SJ coverage",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("log(Minimum coverage of SJ +1)")
} else if (organism_like == 'manatee'){
  p2 <- ggplot(temp_env2$sqanti_data, aes(x=log(min_cov+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette[1])+
  labs(title="Minimum SJ coverage",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("log(Minimum coverage of SJ +1)")
}


ggplotly(p2)
```

#### Coverage Comparison: Canonical vs non-Canonical SJ
```{r coverage comparison, echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

### Coverage comparison: canonical vs non-canonical SJ
select_cols=c("chrom", "strand", "genomic_start_coord", "genomic_end_coord", "junction_category", "start_site_category", 
                "end_site_category", "splice_site", "canonical", "sample_with_cov", "total_coverage")
unique_SJ=unique(temp_env$sqanti_data.junc[,select_cols])

p1 <- ggplot(unique_SJ, aes(x=log(total_coverage + 1), fill=canonical, alpha=0.75))+
  geom_density(na.rm = T) +
  mytheme +
  labs(title="Total coverage of unique SJ",
       subtitle = "Comparison between canonical and non-canonical SJ") +
  xlab("log(Total coverage of SJ +1)")

ggplotly(p1)
```

```{r coverage comparison 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

### Coverage comparison: canonical vs non-canonical SJ file 2
select_cols=c("chrom", "strand", "genomic_start_coord", "genomic_end_coord", "junction_category", "start_site_category", 
                "end_site_category", "splice_site", "canonical", "sample_with_cov", "total_coverage")
unique_SJ=unique(temp_env2$sqanti_data.junc[,select_cols])

p2 <- ggplot(unique_SJ, aes(x=log(total_coverage + 1), fill=canonical, alpha=0.75))+
  geom_density(na.rm = T) +
  mytheme +
  labs(title="Total coverage of unique SJ",
       subtitle = "Comparison between canonical and non-canonical SJ") +
  xlab("log(Total coverage of SJ +1)")

ggplotly(p2)
```

#### BUSCO Completeness Results
```{r BUSCO, echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

### BUSCO completness results
kable(temp_env$busco_results, align = "r")

```

```{r BUSCO 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

### BUSCO completness results file 2
kable(temp_env2$busco_results, align = "r")

```

#### Evaluation of Spike-Ins (SIRVs) 

The following metrics and definitions apply to SIRV transcripts:

- **SIRV transcript:**	Transcript mapping to a SIRV sequence
- **Reference SIRV (rSIRV):**	Lexogen SIRV model
- **True Positive detections (TP):**	rSIRVs identified as RM 
- **Partial True Positive detections (PTP):**	rSIRVs identified as ISM or FSM_non_RM
- **False Negative (FN):**	rSIRVs without FSM or ISM
- **False Positive (FP):**	NIC + NNC + antisense + fusion SIRV_transcripts
- **Sensitivity:**	TP/rSIRVs
- **Precision:**	RM/ SIRV_transcripts
- **Non_redundant Precision:**	TP/ SIRV_transcripts
- **Positive Detection Rate:**	unique(TP+PTP)/rSIRVs
- **False Discovery Rate:**	(SIRV_transcripts - RM)/SIRV_transcripts
- **Redundancy:**	(FSM + ISM)/unique(TP+PTP)

**ATENTION**
If in this chunk of the evaluation all the results are 0, please, check if the reference genome and/or transcriptome used for building your transcript-model  contain information about spike-ins.

```{r Bottom, echo=FALSE}
table <- data.frame(
  'Dataset' = 1,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform,
  'Library Preparation' = library_prep,
  'Data Category' = data_category
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

if(sirv_file != 'NA'){
  kable(temp_env$sirv.results_SIRV, align = "r")    
} 
if(ercc_file != 'NA'){
  kable(temp_env$sirv.results_ERCC, align = "r") 
}
if(sequin_file != 'NA'){
  kable(temp_env$sirv.results_SEQUIN, align = "r")  
}
```

```{r Bottom 2, echo=FALSE}
table <- data.frame(
  'Dataset' = 2,
  'Organism' = organism,
  'Tool' = tool_name,
  'Platform' = platform2,
  'Library Preparation' = library_prep2,
  'Data Category' = data_category2
)

# Render the table as Markdown using kable()
knitr::kable(table, format = "markdown", align = "l")

if(sirv_file2 != 'NA'){
  kable(temp_env2$sirv.results_SIRV, align = "r")    
} 
if(ercc_file2 != 'NA'){
  kable(temp_env2$sirv.results_ERCC, align = "r") 
}
if(sequin_file2 != 'NA'){
  kable(temp_env2$sirv.results_SEQUIN, align = "r")  
}

```

### Benchmarking {.tabset .tabset-fade .tabset-dropdown}

#### Detected Transcripts
```{r Transcripts fig,echo=FALSE}
suppressWarnings(
  if(organism_like == 'mouse'){
  mouse_4a(new_data = c(paste0(platform,'_1'), nrow(temp_env$sqanti_data), 
                sum(temp_env$sqanti_data$structural_category == 'FSM'), 
                sum(temp_env$sqanti_data$structural_category == 'ISM'),
                sum(temp_env$sqanti_data$structural_category == 'NIC'),
                sum(temp_env$sqanti_data$structural_category == 'NNC'),
                sum(temp_env$sqanti_data$structural_category == 'Antisense'),
                sum(temp_env$sqanti_data$structural_category == 'Fusion'),
                sum(temp_env$sqanti_data$structural_category == 'Genic_Genomic'),
                sum(temp_env$sqanti_data$structural_category == 'Genic_Intron'),
                sum(temp_env$sqanti_data$structural_category == 'Intergenic'),
                library_prep,
                platform,
                data_category,
                lab,
                tool_name,
                alias,
                paste0(library_prep, '-', platform),
                paste0(library_prep, '-', data_category),
                paste0(library_prep, '-', platform, '-', data_category)), 
           
           new_data_monoexons = c(sum(temp_env$sqanti_data$structural_category == 'FSM' & temp_env$sqanti_data$exons == 1), 
                          sum(temp_env$sqanti_data$structural_category == 'ISM' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'NIC' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'NNC' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'Antisense' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'Fusion' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'Genic_Genomic' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'Genic_Intron' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$structural_category == 'Intergenic' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env$sqanti_data$exons == 1),
                          platform),
           
           new_data2 = c(paste0(platform2, '_2'), nrow(temp_env2$sqanti_data), 
                sum(temp_env2$sqanti_data$structural_category == 'FSM'), 
                sum(temp_env2$sqanti_data$structural_category == 'ISM'),
                sum(temp_env2$sqanti_data$structural_category == 'NIC'),
                sum(temp_env2$sqanti_data$structural_category == 'NNC'),
                sum(temp_env2$sqanti_data$structural_category == 'Antisense'),
                sum(temp_env2$sqanti_data$structural_category == 'Fusion'),
                sum(temp_env2$sqanti_data$structural_category == 'Genic_Genomic'),
                sum(temp_env2$sqanti_data$structural_category == 'Genic_Intron'),
                sum(temp_env2$sqanti_data$structural_category == 'Intergenic'),
                library_prep2,
                platform2,
                data_category2,
                lab,
                tool_name,
                alias,
                paste0(library_prep2, '-', platform2),
                paste0(library_prep2, '-', data_category2),
                paste0(library_prep2, '-', platform2, '-', data_category2)), 
           
           new_data_monoexons2 = c(sum(temp_env2$sqanti_data$structural_category == 'FSM' & temp_env2$sqanti_data$exons == 1), 
                          sum(temp_env2$sqanti_data$structural_category == 'ISM' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'NIC' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'NNC' & temp_env$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'Antisense' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'Fusion' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'Genic_Genomic' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'Genic_Intron' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$structural_category == 'Intergenic' & temp_env2$sqanti_data$exons == 1),
                          sum(temp_env2$sqanti_data$exons == 1),
                          platform2), 
                          comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
} else if (organism_like == 'manatee'){
  # Generate fig 4b
  append_new_data_4b(new_data = t(as.data.frame(as.numeric(temp_env$transcriptome.results$`Transcriptome without reference`[1:10,1]))),
                     new_data2 = t(as.data.frame(as.numeric(temp_env2$transcriptome.results$`Transcriptome without reference`[1:10,1]))), 
                     comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
}
)


```

#### Length Distribution
```{r Length fig,echo=FALSE}
# Generate fig 4c or 4d
if(organism_like == 'mouse'){
  append_new_data_4c(new_data = temp_env$sqanti_data[,c(1,4)],
                     new_data2 = temp_env2$sqanti_data[,c(1,4)], 
                     comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
} else if(organism_like == 'manatee'){
  append_new_data_4d(new_data = temp_env$sqanti_data[,c(1,4)],
                     new_data2 = temp_env2$sqanti_data[,c(1,4)],
                     comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
}
```

#### Splice Junction Benchmarking
```{r SJ fig,echo=FALSE}
#Generate 4e
append_new_data_4e1(new_data = t(as.data.frame(temp_env$transcriptome.results$`Transcriptome without reference`)),
                    new_data2 = t(as.data.frame(temp_env2$transcriptome.results$`Transcriptome without reference`)),
                    comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
append_new_data_4e3(new_data = t(as.data.frame(temp_env$transcriptome.results$`Transcriptome without reference`)),
                    new_data2 = t(as.data.frame(temp_env2$transcriptome.results$`Transcriptome without reference`)),
                    comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
```

#### BUSCO Metrics
```{r BUSCO fig,echo=FALSE}
suppressWarnings(
  #Generate 4f
if(organism_like == 'manatee'){
  p1 <- append_new_data_4f1_manatee(new_data_manatee = temp_env$busco_results,
                                    new_data_manatee2 = temp_env2$busco_results,
                                    comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
  p2 <- append_new_data_4f2_manatee(new_data_manatee = temp_env$busco_results,
                                    new_data_manatee2 = temp_env2$busco_results,
                                    comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
} else if(organism_like == 'mouse'){
  p1 <- append_new_data_4f1_ES(new_data_ES = temp_env$busco_results,
                               new_data_ES2 = temp_env2$busco_results,
                               comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
  p2 <- append_new_data_4f2_ES(new_data_ES = temp_env$busco_results,
                               new_data_ES2 = temp_env2$busco_results,
                               comparison, bambu, rna_bloom, rnaspades, stringtie2_isoquant)
  }
)

p1
p2


```



